### ISSUE TITLE
Client should ignore dependencies with empty ID

### ISSUE THREAD

**Issue by joelverhagen** [NuGet/Home#5711]
Related server side issue:
https://github.com/NuGet/NuGetGallery/issues/4494

Observe this registration blob:
https://api.nuget.org/v3/registration2-gz/baseline/index.json

Search for `"id": ""` (or `"id":""` if you don't have a JSON formatter in your browser). 

Now look at the .nuspec for this package:
```xml
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<package xmlns="http://schemas.microsoft.com/packaging/2011/10/nuspec.xsd">
  <metadata>
    <id>Baseline</id>
    <version>0.5.0.3</version>
    <title>Baseline</title>
    <authors>Jeremy D. Miller, Brandon Behrens, Andrew Kharlamov</authors>
    <owners>Jeremy D. Miller, Chad Myers, Joshua Arnold, Joshua Flanagan</owners>
    <licenseUrl>https://github.com/JasperFX/Baseline/blob/master/LICENSE.txt</licenseUrl>
    <projectUrl>https://github.com/JasperFX/Baseline</projectUrl>
    <description>Grab bag of generic utilities and extension methods for .Net development</description>
    <dependencies>
      <dependency id="" version="" />
    </dependencies>
  </metadata>
</package>
```

Client fails to install this package since it does not expect that empty string ID. I've tried a variety of scenarios and versions and discovered the following:

1. **Regession**: client 3.5.0 and newer reject this .nuspec (wherever it comes from -- V2, V3, local) due to `ArgumentException` on `id`.
1. **Caused by server bug**: All 3.x and 4.x clients reject registration blobs with this empty ID.

Note that client already gracefully handles empty version range via https://github.com/NuGet/NuGet.Client/commit/dcd728cf9ff33223cead100938a75d35773cd9c3.

_tl;dr Baseline 0.5.0.3 worked prior to 3.5.0 via a V2 source or from a folder/share. This problem should be fixed on both server and client side._

**Comment by emgarten** [comment]
Old packages commonly left out the version property from dependency elements, in older clients this had a special meaning.

As far as I know having an empty id has never been valid. Older versions of NuGet used NuGet.Core.dll for reading v2 feeds and it typically ignored any errors found in packages, but this was considered a bug and it lead to silently accepting bad packages and resolving random versions of dependencies.

Another example is how older clients handled invalid version ranges, ex: 1.A.B or SemVer 2.0.0 version ranges that they could not parse. Exceptions from these were caught and ignored, and >= 0.0.0 was used instead. A project that works fine with a new client would select extremely old versions of a package on an older client, this could potentially bring in unwanted or unsecure dependencies. 

**tl;dr failing fast is better than quietly ignoring bad packages and selecting random dependencies. A lot of similar feed/nuspec validation has been added to the current client libraries that did not exist in NuGet.Core.dll**

**Comment by joelverhagen** [comment]
I thought about this, but there is no valid interpretation of empty IDs. Some forms of "bad" data could have multiple meanings, as you mentioned about invalid versions. If you are in the business of treating empty fields as something else (which we are with versions), there is only one possible interpretation of an empty ID: that is to ignore it. In other words, there is no possible default value of an empty ID so it will effect no users to ignore it. It will only enable a scenario that once worked (pre 3.5.0).

Failing fast makes sense when it puts the system in an unstable/unpredictable/flawed state.

**Comment by emgarten** [comment]
> there is no valid interpretation of empty IDs

It could be interpreted as a valid id equivalent to string.Empty. The local source could contain `.1.0.0.nupkg` with a nuspec that also has `<id></id>` and everything is a match. It could also query a remote feed for the same value and it may use whatever comes back.

Being extremely invalid should not make something valid :smile:

**Comment by mishra14** [comment]
I agree with @emgarten. There may be some benefit in having accepting an empty `id` string. But this can lead to hidden issues for the user which can surface in different forms.
Further, as a package manager, we can have a min bar of requirement from a valid package and an non-empty `id` field is as min as they come.

Maybe, look at other scenarios - do compilers allow identifiers with an empty name? can you save a file without a name?

**Comment by edouardp** [comment]
Is there a final FIX/WONTFIX decision on this?

**Comment by mishra14** [comment]
tagging @rrelyea for triage.

**Comment by emgarten** [comment]
This behavior is by design, the client fails on invalid packages here which seems safer than silently ignoring the problems.

**Comment by edouardp** [comment]
Thanks for making a decision on this.

What is the advice for existing package authors that have already published packages that have invalid specifications in them (e.g. an empty id) back when the tools didn't flag this as an error? Specifically do you have a way forward for the issue in the baseline project that prompted this issue to be opened here: https://github.com/NuGet/NuGetGallery/issues/4494

**Comment by skofman1** [comment]
@edouardp, regarding https://www.nuget.org/packages/Baseline, from what I see the dependency issue was in version 0.5.0.3, and doesn't appear in the latest version (1.4.0), I didn't go over all the versions in the middle. My recommendation would be to unlist all versions that have the issue, so new consumers won't be able to find them any more, and to encourage other consumers to move to the latest version.

**Comment by bchavez** [comment]
@skofman1 Your recommendation does not work. The affected [Baseline](https://www.nuget.org/packages/Baseline) versions have been delisted by the author in https://github.com/JasperFx/baseline/issues/14.

Specifically, the affected `Baseline` versions are: 

* 0.5.0.3
* 0.5.0.4
* 0.5.0.6
* 0.5.0.7

And the problem still persists...

![cmd_714](https://user-images.githubusercontent.com/478118/31753136-aa1c82c6-b443-11e7-8192-75d405dab95d.png)

I'm currently in the process of scanning the entire NuGet package catalog to see if this problem exists in other packages. Almost done, **34,500** packages left to scan. :hourglass_flowing_sand: :mag:  ...
